File: web/scripts/test-p1.mjs
// P1 验收测试：搜索 API（多模式检索/排序/过滤/边界）、点击加权、
// UI 约定（面包屑/导航高亮/骨架屏）、数据完整性。
// 运行：node scripts/test-p1.mjs（需 web 与 data-service 均已启动）

import { PrismaClient } from "@prisma/client";

const BASE = process.env.WEB_URL ?? "http://localhost:3000";
const prisma = new PrismaClient();
const results = [];

function check(name, cond, detail = "") {
  results.push([name, !!cond, detail]);
  console.log(`[${cond ? "PASS" : "FAIL"}] ${name}${cond ? "" : `  <- ${String(detail).slice(0, 180)}`}`);
}

async function search(q, type = "all", limit = 10) {
  const res = await fetch(
    `${BASE}/api/search?q=${encodeURIComponent(q)}&type=${type}&limit=${limit}`,
  );
  return { status: res.status, body: await res.json() };
}

try {
  // ---------- 1. 多模式检索与排序 ----------
  let r = await search("600519");
  check(
    "代码精确检索：600519 首位且基础分满分（100 + 个性化加权 ≤15）",
    r.body.results?.[0]?.code === "600519" &&
      r.body.results[0].score >= 100 &&
      r.body.results[0].score <= 115,
    JSON.stringify(r.body.results?.[0]),
  );

  r = await search("贵州茅台");
  check(
    "名称检索：贵州茅台居首且分数高于其他命中",
    r.body.results?.[0]?.name === "贵州茅台" &&
      r.body.results[0].score > (r.body.results[1]?.score ?? 0),
    JSON.stringify(r.body.results?.slice(0, 3)),
  );

  r = await search("gzmt");
  check(
    "拼音首字母检索命中贵州茅台",
    r.body.results?.some((x) => x.code === "600519"),
    JSON.stringify(r.body.results?.slice(0, 2)),
  );

  r = await search("华夏成长", "fund");
  check(
    "基金名称检索（类别过滤 + 前缀命中）",
    r.body.results?.length > 0 && r.body.results.every((x) => x.type === "fund"),
    JSON.stringify(r.body.results?.slice(0, 2)),
  );

  // ---------- 2. 行情富集 ----------
  r = await search("新能源", "fund", 5);
  check(
    "基金结果行情富集（含价格与涨跌幅）",
    r.body.results?.some((x) => x.price != null && x.changePct != null),
    JSON.stringify(r.body.results?.[0]),
  );

  r = await search("强达转债", "bond", 3);
  check(
    "可转债结果行情富集",
    r.body.results?.some((x) => x.price != null),
    JSON.stringify(r.body.results?.[0]),
  );

  // ---------- 3. 过滤与边界 ----------
  r = await search("茅台", "bond");
  check("类别过滤生效（债券类别下无茅台）", r.status === 200 && r.body.count === 0, JSON.stringify(r.body));

  r = await search("");
  // R14（P2）起：空关键词进入分类浏览模式（browse 结构），不再是"空结果"
  // （2026-09-13 代码审查时更新此过时断言）
  check(
    "空查询进入浏览模式而非报错",
    r.status === 200 && r.body.browse === true && Array.isArray(r.body.items),
    `status=${r.status} body=${JSON.stringify(r.body).slice(0, 120)}`,
  );

  r = await search('茅台" OR "贵州');
  check("FTS 特殊字符不导致 500", r.status === 200, `status=${r.status}`);

  r = await search("zzz不存在的标的999");
  check(
    "无命中查询返回空列表",
    r.status === 200 && Array.isArray(r.body.results) && r.body.count === 0,
    `status=${r.status} count=${r.body.count}`,
  );

  // ---------- 4. 点击加权闭环（含测试数据清理） ----------
  const started = new Date();
  const before = (await search("平安银行", "stock", 1)).body.results[0];
  await fetch(`${BASE}/api/search/click`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ query: "平安银行", type: "stock", code: "000001" }),
  });
  const after = (await search("平安银行", "stock", 1)).body.results[0];
  check("点击一次后排序分 +3", after.score === before.score + 3, `${before.score} -> ${after.score}`);

  const del = await prisma.searchClickLog.deleteMany({
    where: { query: "平安银行", ts: { gte: started } },
  });
  check("测试点击记录已清理", del.count >= 1, `deleted=${del.count}`);

  const badRes = await fetch(`${BASE}/api/search/click`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ query: "x" }),
  });
  check("点击接口缺字段 → 400", badRes.status === 400, `status=${badRes.status}`);

  // ---------- 5. UI 约定（SSR HTML） ----------
  const searchHtml = await (await fetch(`${BASE}/search?q=600519`)).text();
  check("搜索页含面包屑", searchHtml.includes('aria-label="面包屑"'));
  check("顶部导航高亮（aria-current）", searchHtml.includes('aria-current="page"'));

  const productHtml = await (await fetch(`${BASE}/product/stock/600519`)).text();
  check(
    "详情页面包屑含当前标的",
    productHtml.includes('aria-label="面包屑"') && productHtml.includes("贵州茅台"),
  );
  check("详情页流式骨架屏（loading.tsx 生效）", productHtml.includes("animate-pulse"));

  // ---------- 6. 数据完整性 ----------
  const counts = {};
  for (const t of ["stock", "fund", "bond"]) {
    counts[t] = await prisma.product.count({ where: { type: t } });
  }
  check(
    "产品主数据齐备（stock>5000 / fund>20000 / bond>500）",
    counts.stock > 5000 && counts.fund > 20000 && counts.bond > 500,
    JSON.stringify(counts),
  );

  const fts = Number((await prisma.$queryRawUnsafe("SELECT COUNT(*) AS c FROM Product_fts"))[0].c);
  // 口径修正（2026-09-13）：含全部类型（P5 新增 us，如 AAPL）——此前仅算 stock/fund/bond 导致误报
  const allCounts = await prisma.product.groupBy({ by: ["type"], _count: true });
  const total = allCounts.reduce((s, x) => s + Number(x._count), 0);
  check(
    "FTS 索引行数与产品总数一致",
    fts === total,
    `fts=${fts} products=${total}（${allCounts.map((x) => `${x.type}:${x._count}`).join(" ")}）`,
  );

  const withText = await prisma.product.findFirst({
    where: { type: "stock", searchText: { not: null } },
    select: { searchText: true },
  });
  check("searchText 已构建（含二元组/拼音）", (withText?.searchText ?? "").includes(" "));
} finally {
  await prisma.$disconnect();
}

const fails = results.filter((r) => !r[1]);
console.log(`\n===== ${results.length - fails.length}/${results.length} 通过 =====`);
if (fails.length) {
  for (const [name, , detail] of fails) console.log(`  - ${name}: ${String(detail).slice(0, 180)}`);
  process.exit(1);
}

File: data-service/tests/test_p1.py
"""P1 集成测试：产品列表接口 + 行情覆盖（股票/场内基金/场外基金/可转债）。

运行（先启动 uvicorn）：
    cd data-service && PYTHONUTF8=1 .venv/Scripts/python tests/test_p1.py

注意：会触发少量外部请求（东财/天天基金），避免连续高频执行。
"""

import json
import sys

import requests

BASE = "http://localhost:8000"
results: list[tuple[str, bool, str]] = []


def check(name: str, cond, detail: str = "") -> None:
    results.append((name, bool(cond), detail))
    print(f"[{'PASS' if cond else 'FAIL'}] {name}" + (f"  <- {detail}" if not cond and detail else ""))


def get(path: str, **params) -> requests.Response:
    return requests.get(f"{BASE}{path}", params=params, timeout=180)


# ---------- 1. 产品列表 ----------
r = get("/products", type="bond")
body = r.json()
items = body.get("products") or []
check(
    "products?type=bond 返回 200 且数量 > 500",
    r.status_code == 200 and body.get("count", 0) > 500,
    f"status={r.status_code} count={body.get('count')}",
)

sample = items[0] if items else {}
check(
    "产品 schema 字段齐备",
    all(k in sample for k in ("type", "code", "name", "pinyin", "pinyinInitials", "exchange", "tags")),
    json.dumps(sample, ensure_ascii=False)[:180],
)
check(
    "可转债标签正确",
    all("可转债" in (p.get("tags") or []) for p in items[:100]),
    json.dumps(items[0], ensure_ascii=False)[:180],
)
check(
    "代码与名称无空值（抽样 200）",
    all(p["code"] and p["name"] for p in items[:200]),
)

r = get("/products", type="crypto")
# crypto 已有 CoinGecko 列表源（crypto_provider register_list）。可达时返回 200 且有数据，
# 不可达（R12 代理不可用）时返回 502 并显式降级——两者皆正确，不能写死 502。
_body = (r.json() or {}) if r.status_code == 200 else {}
check(
    "products?type=crypto 可用或显式降级（200+数据 或 502）",
    (r.status_code == 200 and len(_body.get("products") or []) > 0)
    or (r.status_code == 502),
    f"status={r.status_code} n={len(_body.get('products') or [])} body={r.text[:120]}",
)

r = get("/products", type="option")
check("products 未支持类型 → 400", r.status_code == 400, f"status={r.status_code}")

# ---------- 2. 批量行情覆盖 ----------
r = get("/quotes", type="stock", codes="600519,000001")
qs = (r.json() or {}).get("quotes") or {}
check(
    "股票批量行情 2/2 且含价格",
    len(qs) == 2 and all(q.get("price") for q in qs.values()),
    json.dumps(qs, ensure_ascii=False)[:180],
)

r = get("/quotes", type="fund", codes="158008")
q = ((r.json() or {}).get("quotes") or {}).get("158008") or {}
check(
    "场内基金（ETF）实时行情含涨跌幅",
    bool(q.get("price")) and q.get("changePct") is not None,
    json.dumps(q, ensure_ascii=False)[:180],
)

r = get("/quotes", type="fund", codes="000001")
q = ((r.json() or {}).get("quotes") or {}).get("000001") or {}
check(
    "场外基金每日净值（单位净值 + 日增长率）",
    bool(q.get("price")) and q.get("changePct") is not None and "nav" in str(q.get("source", "")),
    json.dumps(q, ensure_ascii=False)[:180],
)

# 转债行情：标的改为活跃代码段候选池（未上市/已退市转债无行情属正常，
# 2026-09-13 code review：原写死 123284 一旦退市即误报）
r = get("/quotes", type="bond", codes="123284,123283,123282,111000,113050")
_qs = (r.json() or {}).get("quotes") or {}
_hit = next((k for k, v in _qs.items() if v.get("price")), "")
check(
    "可转债实时行情可用（活跃候选任一命中）",
    bool(_hit),
    f"quotes={list(_qs.keys())} body={r.text[:150]}",
)

# ---------- 3. 单只查询（详情页用） ----------
r = get("/quote", type="fund", code="000001")
check(
    "单只场外基金 quote 可用",
    r.status_code == 200 and bool((r.json() or {}).get("price")),
    r.text[:150],
)

# ---------- 4. 边界 ----------
r = requests.get(f"{BASE}/quotes", params={"type": "stock", "codes": ""}, timeout=30)
check("quotes 空 codes → 400", r.status_code == 400, f"status={r.status_code}")

r = get("/quotes", type="option", codes="600519")
check("quotes 未支持类型 → 400", r.status_code == 400, f"status={r.status_code}")

# ---------- 汇总 ----------
fails = [x for x in results if not x[1]]
print(f"\n===== {len(results) - len(fails)}/{len(results)} 通过 =====")
if fails:
    print("失败项：")
    for name, _, detail in fails:
        print(f"  - {name}: {detail[:180]}")
sys.exit(1 if fails else 0)
